### ISSUE TITLE
Add support for tikz and tikzcd

### ISSUE THREAD

**Issue by lusi8559** [KaTeX/KaTeX#219]
Hi,

Could you add support for tikz and tikzcd so diagrams such as

![image](https://cloud.githubusercontent.com/assets/9576624/7102953/1e0dc886-e050-11e4-8d25-7fcb832975e5.png)

![image](https://cloud.githubusercontent.com/assets/9576624/7102956/2d62316e-e050-11e4-8b7e-8a1b5040a819.png)

![image](https://cloud.githubusercontent.com/assets/9576624/7102958/45043b3c-e050-11e4-97d5-96fd09704675.png)

can be supported?

**Comment by spicyj** [comment]
This would be cool, but honestly, tikz support would be a gargantuan undertaking, probably larger than all of KaTeX itself at present, so we're unlikely to do this. You could use SVG to draw similar diagrams on the web.

**Comment by tobiasdiez** [comment]
Is there at least any hope that at least commutative diagrams in the spirit of tikzcd could be supported in the future? Similar to the [amscd extension](http://docs.mathjax.org/en/latest/input/tex/extensions/amsCd.html) for MathJax just using tikzcd syntax (and of course for KaTeX). At least basic support of commutative diagrams would be nice.

**Comment by kevinbarabash** [comment]
This feels like something that should be done as an extension, similar to our `mhchem` support.

**Comment by edemaine** [comment]
I'm going to reopen this as it's a feature request I hear often, so it'd be nice if it got some attention.

**Comment by jxxcarlson** [comment]
Hello!  I am the developer of [minilatex.lamdera.app](https://minilatex.lamdera.app) which has been using KaTeX as of this past August.  The project has matured since then, and I have a group of math profs who want to use it for a collaborative project.  They will no doubt use commutative diagrams, which is supported by the amscd package.  Has there been any progress on this (as an extension perhaps?).  If not perhaps there is a way I could help out to expedite this

**Comment by edemaine** [comment]
@jxxcarlson No progress that I'm aware of. You're welcome to try writing an extension, probably starting with a small subset to target. But as pointed out above, it's likely a large undertaking.  I also wonder whether it's possible without knowing the width/height of LaTeX renderings, which KaTeX currently doesn't support (specifically width is not known/computed, in particular because font ligatures are complicated).

It looks like [quiver](https://github.com/varkor/quiver) is an interactive GUI web-based commutative diagram editor that uses KaTeX to render math/labels, and uses SVG to render the arrows (perhaps based on the way KaTeX does). Maybe there's a way to use that as an alternative?

**Comment by jxxcarlson** [comment]
Thanks very much for your reply.  I will investigate quiver first.

Sadly, commutative diagrams are an essential part of many fields of mathematics, including mine (algebraic geometry).  To that end, would implementing something like the AmsCD package in KaTeX be easier than tikzcd?

I had been using MathJax in my project, but since it is based on Elm,, math rendering needs to be done via a custom element. The KaTeX

	katex.renderToString

Is very performant in this respect.  Not quite sure why the big performance difference.  Perhaps MathJax was sets up its entire environment for each element.  Dunno.

Best regards,

Jim Carlson

**Comment by edemaine** [comment]
As an update to this, AMS's `CD` environment is now supported in the master branch, thanks to #2396. So you should give that a try! (e.g. on our demo page) Hopefully this suffices for a many people's commutative diagrams, as tikz would be very complicated to implement.

@jxxcarlson KaTeX is generally much faster than MathJax. This used to be largely because KaTeX doesn't query the DOM while MathJax doesn't, but I'm not sure if that's the only reason or even if it's still true of MathJax.

**Comment by NickHu** [comment]
Here's a radical idea, which is mostly already done: use one of the ports of pdflatex to webassembly. Someone even already wrote one for tikz here: https://tikzjax.com/. KaTeX could detect tikzpicture environments and then pass them to this.

**Comment by edemaine** [comment]
I had seen this before but hadn't thought of it as a fallback for e.g. tikz. Perhaps a KaTeX plugin could be written to do so. (I don't think it makes sense as part of base KaTeX, as the wasm build is quite large.)

**Comment by honeycombi** [comment]
Is there any updates for this? I'm not sure how related https://github.com/artisticat1/obsidian-tikzjax is to this, but it might be a head start? AMSCD is very limited (doesn't support loops back the same point as far as I know).

**Comment by edemaine** [comment]
Obsidian TikZJax uses the TikZJax project mentioned above, which runs actual LaTeX directly via WebAssembly. I don't think it's within the scope of KaTeX, but could be an alternative to (and possibly a fallback for) KaTeX.

**Comment by Speculate7348** [comment]
This needs to be supported.

**Comment by Speculate7348** [comment]
No update?

**Comment by yosun** [comment]
indeed

**Comment by jeffreyschultz** [comment]
Any plans to implement this feature request?

**Comment by yosun** [comment]
i think you can probably vibe code it nowadays

**Comment by Speculate7348** [comment]
> i think you can probably vibe code it nowadays

Cringe take
